### ISSUE TITLE
Preserve symlinks

### ISSUE THREAD

**Issue by Globegitter** [pytest-dev/pytest#5266]
It seems over the time pytest introduced `realpath` calls to resolve symlinks. When trying to use pytest with [https://bazel.build/](bazel) however that falls flat because is executing all tests in an isolated sandbox (think lightweight docker) and is for performance reason just symlinking files rather than copying them and it expects tools to respect symlinks. Further it only puts the files into the sandbox that it needs for that test, so it might just be the pytest dependency, the testfile itself and the library that is needed for it. The problem I am now facing is, that pytest is resolving the symlink of the testfile, ending up in my real project directory, discovering a `__init__.py` file there and then complaining about missing imports because the isolated python provided by bazel does not have access to these for running the test.

It would be nice to get an option for pytest to just treat symlinks as the real files and not try to resolve them. In the nodejs world for example the flag `--preserve-symlinks` exists for that and it would be great to get the same for pytest.

Let me know if you need a minimal reproduction case, then I am happy to provide one.

pytest version 4.4.2, OS: Elementary OS 5 (Ubuntu 18.04)

**Comment by blueyed** [comment]
> `--preserve-symlinks` exists for that

Sounds sensible.

Is there another way that wouldn't require a flag?

IIRC I've started with that to make `pytest tests/…` work with `tests` being a symlink to `project/foo/tests`.
After all full resolving is not needed here maybe, but only for tests etc.

A minimal example would be great to have.

**Comment by Globegitter** [comment]
The way bazel executes `pytest` in our case is by passing it all the filepaths relative from the working directory, so e.g. `pytest path/to/test.py` or if there would be multiple ones `pytest path/to/test1.py path/to/test2.py`. So possibly symlinks could be preserved only in the case of passing it direct file paths. WDYT?

But yeah will work on a repro asap.

**Comment by Globegitter** [comment]
Here is a minimal example: https://github.com/Globegitter/pytest-repro

**Comment by blueyed** [comment]
> So possibly symlinks could be preserved only in the case of passing it direct (relative) file paths. WDYT?

Might be a good compromise.

But it is also about internal lookup of contests etc (from parent dirs), and that would then mean to still resolve them internally at least in those places.
IIRC I've tried to adhere to this back then.

It happens here: https://github.com/pytest-dev/pytest/blob/c8f7e50c47cc771142f2681aebf7dd9f86c83853/src/_pytest/main.py#L710
It was done for https://github.com/pytest-dev/pytest/issues/4325 via https://github.com/pytest-dev/pytest/pull/4337.

**Comment by blueyed** [comment]
Started working on it in #5270 (8 failures IIRC).
Moved your repro into a test: https://github.com/pytest-dev/pytest/pull/5270/files#diff-f31e02762db03e39bb7fe5f9fb52eb44 - does it match your expectations?

Some question came up already: how should nodes be reported?  Using the symlink or target?  (https://github.com/pytest-dev/pytest/blob/c8f7e50c47cc771142f2681aebf7dd9f86c83853/testing/test_collection.py#L1134)

**Comment by Globegitter** [comment]
Thanks for the quick progress! Just tested it out and it does indeed work as expected - great. When you say "how should nodes be reported", do you mean which path it should report, when it prints to stdout?
E.g.
```
collected 1 item

path/to/symlink/test.py .                                    [100%]
```

I think in this case either or is fine, but I would prefer the symlink path, the behaviour it seems to be now as I can tell. That would imo also fit the behaviour of "preserve symlinks".

**Comment by Globegitter** [comment]
@blueyed is there anything I can do to push this forward? Happy to also put in some time myself if I know what the best way forward is.

**Comment by blueyed** [comment]
@Globegitter 
> do you mean which path it should report, when it prints to stdout?

Yes - basically how the nodeid should look like (that gets used internally also, including for test selection).
I agree that having the symlink there is more useful.

You could pick up my branch and make the failing tests work there, I assume that the resolving has to be done in a more inner place now - possibly only temporarily.

**Comment by blueyed** [comment]
@Globegitter 
Any chance you had worked on / looked into it?

**Comment by chaokunyang** [comment]
I came across the same problem.
For example, when I execute `pytest ray/streaming/test/streaming_cluster_test.py`, the module name changed from `ray.streaming.test.streaming_cluster_test` to `python.test.streaming_cluster_test`. Because I symlinked  `ray/streaming/test/` to `python/streaming/test`.
Would this be fixed or is that I use pytest in wrong way? I must use symlinks.  So if this won't be fixed, I'll have to write w  script to import my code and test it, which will be less good

**Comment by aditya-qure** [comment]
Any updates on this? I am getting 0% coverage.

Thanks

**Comment by nicoddemus** [comment]
We have since then changed pytest and I think we do respect symlinks now: we no longer use `Path.resolve()`, using `Path.absolute()` instead when needed... there might be someplace where this is still an issue thought that we missed.

@aditya-qure can you post a MWE?

**Comment by Zac-HD** [comment]
Closing as resolved: pytest stopped force-resolving symlinks during collection back in pytest 6.0.0 via #6523 (fixing #5965), switching from `Path.resolve()` to `Path.absolute()`.
